base/base32768_body.hpp

```diff
@@ -241,7 +262,47 @@ UniqueArray<std::uint8_t> Base32768Encode(Array<std::uint8_t const> input,
 void Base32768Decode(Array<std::uint8_t const> input,
                      Array<std::uint8_t> output) {
   CHECK_NOTNULL(input.data);
-  CHECK_NOTNULL(output.data);
+  CHECK(input.size == 0 || output.data != nullptr);
+
+  std::uint8_t const* const input_end = input.data + input.size;
+  std::uint8_t const* const output_end = output.data + output.size;
+  std::int64_t output_bit_index = 0;
+  while (input.data < input_end) {
+    bool const at_end = input_end - input.data == 2;
+    char16_t code_point;
+    std::memcpy(&code_point, input.data, sizeof(char16_t));
+    std::int32_t data;
+    std::int32_t shift = bytes_per_code_point * bits_per_byte -
+                         bits_per_code_point - output_bit_index;
+    Repertoire const* repertoire = &fifteen_bits;
+    if (at_end && seven_bits.CanEncode(code_point)) {
+      shift = bytes_per_code_point * bits_per_byte - bits_per_final_code_point -
+              output_bit_index;
+      repertoire = &seven_bits;
+    }
+
+    // Align |data| on the output bit index.
+    data = repertoire->Decode(code_point);
+    data <<= shift;
+
+    // Fill the output with the parts of the code point belonging to each byte.
+    if (output_bit_index == 0) {
+      output.data[0] = (data >> (2 * bits_per_byte));
```

Comment:
Could we just 0 `output` at the beginning to simplify the logic here?

Reply:
As discussed, not easily.